tests/test_qubit_device.py

```diff
@@ -740,257 +550,27 @@ def test_non_analytic_var(self, mock_qubit_device_with_original_statistics, monk
-        with monkeypatch.context() as m:
-            m.setattr(QubitDevice, 'rotate_basis', lambda self, op: None)
-            m.setattr(QubitDevice, 'generate_samples', lambda self: times_called.append(1))
-            m.setattr(QubitDevice, 'probability', lambda self, wires=None: [1,1])
-            m.setattr(QubitDevice, 'pauli_eigvals_as_samples', lambda self, wires: pauli_eigvals_call_history.append(1))
-            res = mock_qubit_device_with_original_statistics.sample(obs)
-
-        assert pauli_eigvals_call_history == [1]
-
-    def test_pauli_eigvals_called_rotated_prob_not_none(self, mock_qubit_device_with_original_statistics, monkeypatch):
-        """Tests that the sample method calls the pauli_eigvals_as_samples when the _rotated_prob attribute is  None
-
-        Additional QubitDevice methods that are mocked:
-        -rotate_basis
-        -probability
-        -pauli_eigvals_as_samples
-        """
-        assert mock_qubit_device_with_original_statistics._rotated_prob is None
-
-        # Setting the attribute to not None
-        mock_qubit_device_with_original_statistics._rotated_prob = 'NotNone'
-
-        obs = qml.PauliX(0)
-
-        pauli_eigvals_call_history = []
-        with monkeypatch.context() as m:
-            m.setattr(QubitDevice, 'rotate_basis', lambda self, op: None)
-            m.setattr(QubitDevice, 'generate_samples', lambda self: times_called.append(1))
-            m.setattr(QubitDevice, 'probability', lambda self, wires=None: [1,1])
-            m.setattr(QubitDevice, 'pauli_eigvals_as_samples', lambda self, wires: pauli_eigvals_call_history.append(1))
-            res = mock_qubit_device_with_original_statistics.sample(obs)
-
-        assert pauli_eigvals_call_history == [1]
-
-    def test_custom_eigvals_called_rotated_prob_none(self, mock_qubit_device_with_original_statistics, monkeypatch):
-        """Tests that the sample method calls the custom_eigvals_as_samples when the _rotated_prob attribute is  None
-
-        Additional QubitDevice methods that are mocked:
-        -rotate_basis
-        -probability
-        -pauli_eigvals_as_samples
-        """
-        assert mock_qubit_device_with_original_statistics._rotated_prob is None
-
-        obs = qml.PauliX(0) @ qml.PauliX(1)
-
-        custom_eigvals_call_history = []
-        with monkeypatch.context() as m:
-            m.setattr(QubitDevice, 'rotate_basis', lambda self, op: None)
-            m.setattr(QubitDevice, 'generate_samples', lambda self: times_called.append(1))
-            m.setattr(QubitDevice, 'probability', lambda self, wires=None: [1,1])
-            m.setattr(QubitDevice, 'custom_eigvals_as_samples', lambda self, wires, eigvals: custom_eigvals_call_history.append(1))
-            res = mock_qubit_device_with_original_statistics.sample(obs)
-
-        assert custom_eigvals_call_history == [1]
-
-    def test_custom_eigvals_called_rotated_prob_not_none(self, mock_qubit_device_with_original_statistics, monkeypatch):
```

Comment:
Can you safely remove all of those tests? I guess some functionality they tested is still there?

Reply:
(For these parts, I was helping out a bit.)

This version of ``_qubit_device`` reverted certain reorganizations of the code (e.g. the ``pauli_eigvals_as_samples`` and the ``custom_eigvals_as_samples`` methods have been removed and the original code was replaced), therefore certain tests that were testing branches of inside ``sample`` could be removed. In specific it would go something along the lines of:
``test_rotated_prob_none_rotate_with_memory``: method calls do not take place anymore
``test_rotated_prob_none_rotate_basis_without_memory``: method calls do not take place anymore
``test_rotated_prob_not_none_no_rotate_basis``: ``_rotated_prob`` attribute is not used anymore
``test_pauli_eigvals_called_rotated_prob_none``: ``pauli_eigvals_as_samples`` has been removed
``test_pauli_eigvals_called_rotated_prob_not_none``: ``pauli_eigvals_as_samples`` has been removed
``test_custom_eigvals_called_rotated_prob_none``: ``custom_eigvals_as_samples`` and the ``_rotated_prob`` attribute have been removed
``test_custom_eigvals_called_rotated_prob_not_none``: ``pauli_eigvals_as_samples`` and the ``_rotated_prob`` attribute have been removed

This doesn't mean that there wouldn't be any functionality that could be tested anyways. Would you think that there was something that could still be kept?

Reply:
I think I'm not in a place to decide if some of those tests should be kept (in functional terms), because my overview of the new structure is not so good. I was just amazed that we could remove so many lines of test code

Reply:
Yes, there are quite some changes going on in the structure pretty rapidly :sweat_smile: It is a totally good point, that lots were removed and valuable to have another look if something worthy was actually deleted.

The previous (current master) version contained the following ``sample`` method:
```python
    def sample(self, observable):
        wires = observable.wires
        name = observable.name

        # TODO: remove the following part as we can assume that
        # rotate_basis has already been called. Doing so will, however break
        # unit tests directly calling on sample()
        if self._rotated_prob is None:
            observable.return_type = Sample
            self.rotate_basis([observable])

            if self._memory or (not self.analytic):
                self.generate_samples()

        if isinstance(name, str) and name in {"PauliX", "PauliY", "PauliZ", "Hadamard"}:
            return self.pauli_eigvals_as_samples(wires)

        # Need to post-process the samples using the observables.
        return self.custom_eigvals_as_samples(wires, observable.eigvals)
```
Most tests were cases for each branch of the if statements.